src/posts/abstracting-the-frontend/index.md

```diff
@@ -0,0 +1,124 @@
+
+It is my opinion that today's front-end developer should harness the skill to provide a user with a visual interface
+to interact with the product one is working on. Be it a web application, a mobile app, a car's dashboard or a kitchen fridge.
+This is inherently the job of the front-end developer.
+
+We can safely assume that the front end can appear in many forms, sizes and places, and all these require their own unique
+behaviour in order to fulfill the needs of our users.
+
+As an organisation grows, it is almost certain that the product it provides will need to be presented in more than one of these
+types of platforms. For example, a media streaming service will effectively want to be present in all of them,
+perhaps even the fridge!
+
+### Separation of concerns
+
+Software architects often like to use layers to sort the different responsibilities of a program into distinct,
+easy-to-reason-about chunks that can be stacked to form a hierarchy. I will attempt to do the same thing with what I believe
+to be the chunks that a front end should consist of, so that we can maintain the same level of flexibility and productivity
+across all the types of gadgets and gizmos that exist out there.
+
+_**Quick disclaimer:** Perhaps these concepts can be sorted into something more pragmatic than layers,
+the important part is the separation of concerns._
+
+![Front End Layers](./diagram.png)
+_An illustration of the Front End Layers_
+
+#### Design system
+
+As front-end developers, UI and UX designers we are responsible for the visual representation and interaction of our products.
+What we present to our users should always be the number one concern of any front-end unit. Building a design system early
+and sticking to it as if it was holy law is imperative to succeed in scaling your products.
+The implications of what a design system is may vary but these are the assertions that I believe we should make about them.
+
+1. The core of a design system should be reusable across products. A company's products don't have to look the same,
+   but they should feel the same.
+2. Component libraries should exist that strictly follow the rules of the design system.
+3. A component library should not consume remote data directly.
+4. A design system's documentation should be available to the public.
+
+**Note:** Point 3 makes the layer approach a bit awkward, but I hope we can agree on the general philosophy.
+
+#### Data aggregate
+
+If we ever assumed that the backend team is only concerned delivering data to our front ends, this is no longer the case.
+Backend developers building API's have their own group of stakeholders to worry about, integrators. This means that there will be
+certain compromises in API design. Thus, having a middleware where we can create new datasets tailored
+to the front end can prove quite useful.
+
+A front end that does not communicate with a data source is a rare sight these days and I believe that the number of data points a
```

Comment:
At some point in recent computer history you would have said "A desktop client that doesn't communicate with a data source is a rare sight these days and..."